crates/core/c8y_api/src/smartrest/operations.rs

```diff
@@ -10,18 +29,326 @@ use std::collections::HashSet;
 use std::fs;
 use std::path::Path;
 use std::path::PathBuf;
+use std::sync::Arc;
 use std::time::Duration;
 use tedge_api::substitution::Record;
 use tedge_config::TopicPrefix;
+use tedge_utils::file;
 use tracing::error;
 use tracing::warn;
 
 use super::payload::SmartrestPayload;
+use super::topic::C8yTopic;
 
 const DEFAULT_GRACEFUL_TIMEOUT: Duration = Duration::from_secs(3600);
 const DEFAULT_FORCEFUL_TIMEOUT: Duration = Duration::from_secs(60);
 
+type ExternalId = String;
+type ExternalIdRef = str;
+
 type OperationName = String;
+type OperationNameRef = str;
+
+/// Used to hold and query supported operations for all devices.
+pub struct SupportedOperations {
+    /// External ID of the main device.
+    ///
+    /// Required because when given an external id of the main device when creating an operation, we want to create it
+    /// in a main directory, instead of a subdirectory.
+    pub device_id: String,
+
+    /// Base c8y operations directory for all devices.
+    ///
+    /// By default `/etc/tedge/operations/c8y`. Contains operation files for the main device, operation templates, and
+    /// directories with operation files for child devices.
+    pub base_ops_dir: Arc<Utf8Path>,
+
+    /// Currently loaded set of supported operations for all registered devices by its external id.
+    pub operations_by_xid: HashMap<String, Operations>,
+}
+
+impl SupportedOperations {
+    /// Add an operation to the supported operation set for a given device.
+    ///
+    /// Creates and writes new operation file to the filesystem.
+    pub fn add_operation(
+        &self,
+        device_xid: &ExternalIdRef,
+        c8y_operation_name: &OperationNameRef,
+    ) -> Result<(), anyhow::Error> {
+        let ops_file = if device_xid == self.device_id {
+            self.base_ops_dir.join(c8y_operation_name)
+        } else {
+            self.base_ops_dir.join(device_xid).join(c8y_operation_name)
+        };
+
+        // Create directory for a device if it doesn't exist yet
+        file::create_directory_with_defaults(ops_file.parent().expect("should never fail"))?;
```

Comment:
Assuming that this will never fail seems a bit too bold here.

Reply:
Above, we add at least 1 component to the `base_ops_dir` path to create `ops_file`, so I'm pretty sure taking the `.parent()` here never fails.

Reply:
You are correct.